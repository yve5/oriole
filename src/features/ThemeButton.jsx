import React, { useEffect, useState } from 'react';

import { Sun, Moon, CircleHalf } from './icons';
import {
  BOOTSTRAP_MODES,
  BOOTSTRAP_MODE_PROP,
  BOOTSTRAP_MODE_SESSION,
} from '../constants';

const ThemeButton = () => {
  const [themeMode, setThemeMode] = useState(0);
  const [autoTheme, setAutoTheme] = useState(0);

  useEffect(() => {
    const savedTheme = Number(sessionStorage.getItem(BOOTSTRAP_MODE_SESSION));
    if (Number.isInteger(savedTheme)) setThemeMode(savedTheme);
  }, []);

  const isDark =
    window.matchMedia &&
    window.matchMedia('(prefers-color-scheme: dark)').matches;

  useEffect(() => {
    const applySystemTheme = () => setAutoTheme(isDark ? 1 : 0);

    applySystemTheme();

    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    mediaQuery.addEventListener('change', applySystemTheme);

    return () => {
      mediaQuery.removeEventListener('change', applySystemTheme);
    };
  }, []);

  useEffect(() => {
    const htmlElement = document.documentElement;
    let mode = themeMode;

    if (mode === 2) {
      mode = autoTheme;
    }

    htmlElement.setAttribute(BOOTSTRAP_MODE_PROP, BOOTSTRAP_MODES[mode]);

    return () => {
      htmlElement.removeAttribute(BOOTSTRAP_MODE_PROP);
    };
  }, [themeMode, isDark]);

  let themeIcon = <Sun />;

  if (themeMode === 1) {
    themeIcon = <Moon />;
  } else if (themeMode === 2) {
    themeIcon = <CircleHalf />;
  }

  return (
    <button
      type="button"
      title={BOOTSTRAP_MODES[themeMode]}
      className="nav-link text-secondary text-white"
      onClick={() => {
        const newMode = (themeMode + 1) % 3;
        sessionStorage.setItem(BOOTSTRAP_MODE_SESSION, newMode);
        setThemeMode(newMode);
      }}
    >
      {themeIcon}
      {/* {autoTheme} */}
    </button>
  );
};

export default ThemeButton;
