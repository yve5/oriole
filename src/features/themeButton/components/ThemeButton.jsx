import React, { useEffect, useState } from 'react';

import { Sun, Moon, CircleHalf } from '../../icons';
import {
  B7P_MODES,
  B7P_MODE_SCHEME,
  B7P_MODE_SESSION,
} from '../resources/constants';

const ThemeButton = () => {
  const [theme, setTheme] = useState(0);
  const [autoTheme, setAutoTheme] = useState(0);

  useEffect(() => {
    const savedTheme = Number(sessionStorage.getItem(B7P_MODE_SESSION));
    if (Number.isInteger(savedTheme)) setTheme(savedTheme);

    const applySystemTheme = () => {
      setAutoTheme(window.matchMedia(B7P_MODE_SCHEME).matches ? 1 : 0);
    };

    applySystemTheme();

    const mediaQuery = window.matchMedia(B7P_MODE_SCHEME);
    mediaQuery.addEventListener('change', applySystemTheme);

    return () => {
      mediaQuery.removeEventListener('change', applySystemTheme);
    };
  }, []);

  useEffect(() => {
    document.documentElement.setAttribute(
      'data-bs-theme',
      B7P_MODES[theme === 2 ? autoTheme : theme]
    );
  }, [theme, autoTheme]);

  let themeIcon = <Sun />;
  if (theme === 1) {
    themeIcon = <Moon />;
  } else if (theme === 2) {
    themeIcon = <CircleHalf />;
  }

  return (
    <button
      type="button"
      title={B7P_MODES[theme]}
      data-testid="theme-button"
      className="nav-link text-secondary text-white"
      onClick={() => {
        const newTheme = (theme + 1) % 3;
        sessionStorage.setItem(B7P_MODE_SESSION, newTheme);
        setTheme(newTheme);
      }}
    >
      {themeIcon}
    </button>
  );
};

export default ThemeButton;
