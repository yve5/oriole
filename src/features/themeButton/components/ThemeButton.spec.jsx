import React from 'react';
import { vi } from 'vitest';
import { act, fireEvent, render } from '@testing-library/react';

import ThemeButton from './ThemeButton';

const mockMatchMedia = (matches) => ({
  removeEventListener: vi.fn(),
  addEventListener: vi.fn(),
  dispatchEvent: vi.fn(),
  matches,
});

describe('ThemeButton', () => {
  it('renders the component. default', async () => {
    window.matchMedia = vi.fn().mockImplementation(() => mockMatchMedia(false));

    const { asFragment, getByTestId, rerender } = render(<ThemeButton />);

    rerender(<ThemeButton />);

    await act(async () => {
      fireEvent.click(getByTestId('theme-button'));
    });

    rerender(<ThemeButton />);

    await act(async () => {
      fireEvent.click(getByTestId('theme-button'));
    });

    rerender(<ThemeButton />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component. dark', async () => {
    window.matchMedia.mockImplementation(() => mockMatchMedia(true));

    const { asFragment, getByTestId, rerender } = render(<ThemeButton />);

    rerender(<ThemeButton />);

    await act(async () => {
      fireEvent.click(getByTestId('theme-button'));
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
