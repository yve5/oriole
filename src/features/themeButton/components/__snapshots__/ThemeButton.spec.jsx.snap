// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeButton > renders the component. dark 1`] = `
<DocumentFragment>
  <button
    class="nav-link text-secondary text-white"
    data-testid="theme-button"
    title="light"
    type="button"
  >
    <svg
      class="bi bi-sun"
      fill="currentColor"
      height="20"
      viewBox="0 0 16 16"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 11a3 3 0 1 1 0-6 3 3 0 0 1 0 6m0 1a4 4 0 1 0 0-8 4 4 0 0 0 0 8M8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0m0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13m8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5M3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8m10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0m-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0m9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707M4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`ThemeButton > renders the component. default 1`] = `
<DocumentFragment>
  <button
    class="nav-link text-secondary text-white"
    data-testid="theme-button"
    title="auto"
    type="button"
  >
    <svg
      class="bi bi-circle-half"
      fill="currentColor"
      height="20"
      viewBox="0 0 16 16"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 15A7 7 0 1 0 8 1zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16"
      />
    </svg>
  </button>
</DocumentFragment>
`;
